src/librustc_mir/transform/promote_consts.rs

```diff
@@ -61,11 +61,15 @@ impl<'tcx> MirPass<'tcx> for PromoteTemps<'tcx> {
 
         let def_id = src.def_id();
 
-        let mut rpo = traversal::reverse_postorder(body);
-        let (temps, all_candidates) = collect_temps_and_candidates(tcx, body, &mut rpo);
+        let read_only_body = read_only!(body);
```

Comment:
What kind of strange hack is this?

Reply:
(I found the `read_only!` macro definition but it seems undocumented.)

Reply:
unrelated to this PR, but yea, this can likely become a method. I think the reason it's a macro is multiple refactorings after each other removing the need for the macro at some point

Reply:
I think it is a macro because it has a signature like `fn(&mut self) -> &Self`, which runs into a shortcoming of the borrow checker where the returned reference is not allowed to alias.

Reply:
Well that would be a useful thing to have in a doc comment. ;)